pkg/controllers/machinesync/machine_sync_controller.go

```diff
@@ -707,95 +709,142 @@ func (r *MachineSyncReconciler) createOrUpdateCAPIInfraMachine(ctx context.Conte
+	logger.Info("Changes detected for CAPI machine. Updating it", "diff", fmt.Sprintf("%+v", capiMachinesDiff))
 
-	if err := r.Update(ctx, newCAPIMachine); err != nil {
-		logger.Error(err, "Failed to update Cluster API machine")
+	if err := r.Update(ctx, updatedOrCreatedCAPIMachine); err != nil {
+		logger.Error(err, "Failed to update CAPI machine")
 
-		updateErr := fmt.Errorf("failed to update Cluster API machine: %w", err)
+		updateErr := fmt.Errorf("failed to update CAPI machine: %w", err)
 
 		if condErr := r.applySynchronizedConditionWithPatch(ctx, mapiMachine, corev1.ConditionFalse, reasonFailedToUpdateCAPIMachine, updateErr.Error(), nil); condErr != nil {
-			return ctrl.Result{}, utilerrors.NewAggregate([]error{updateErr, condErr})
+			return false, utilerrors.NewAggregate([]error{updateErr, condErr})
 		}
 
-		return ctrl.Result{}, updateErr
+		return false, updateErr
 	}
 
-	logger.Info("Successfully updated Cluster API machine")
-
-	return ctrl.Result{}, nil
+	return true, nil
 }
 
-// createOrUpdateMAPIMachine creates a MAPI machine from a CAPI one, or updates
-// if it exists and it is out of date.
-func (r *MachineSyncReconciler) createOrUpdateMAPIMachine(ctx context.Context, mapiMachine *machinev1beta1.Machine, newMAPIMachine *machinev1beta1.Machine) (ctrl.Result, error) { //nolint:unparam
+// createOrUpdateCAPIMachine creates or updates (if existing but out of date) a CAPI machine from a convertedCAPIMachine (CAPI machine object converted from MAPI).
+//
+//nolint:unparam
+func (r *MachineSyncReconciler) createOrUpdateCAPIMachine(ctx context.Context, sourceMAPIMachine *machinev1beta1.Machine, existingCAPIMachine *clusterv1.Machine, convertedCAPIMachine *clusterv1.Machine) (ctrl.Result, error) {
 	logger := log.FromContext(ctx)
 
-	if mapiMachine == nil {
-		if err := r.Create(ctx, newMAPIMachine); err != nil {
-			logger.Error(err, "Failed to create Machine API machine")
-			return ctrl.Result{}, fmt.Errorf("failed to create Machine API machine: %w", err)
-		}
+	// If there is no existing CAPI machine, create a new one.
+	if created, err := r.ensureCAPIMachine(ctx, sourceMAPIMachine, existingCAPIMachine, convertedCAPIMachine); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to ensure CAPI machine: %w", err)
+	} else if created {
+		// The CAPI machine was created here, return early to allow for the change to be propagated.
+		return ctrl.Result{}, nil
+	}
+	// There already is an existing CAPI machine, work out if it needs updating.
 
-		logger.Info("Successfully created Machine API machine")
+	// Compare the existing CAPI machine with the converted CAPI machine to check for changes.
+	capiMachinesDiff := compareCAPIMachines(existingCAPIMachine, convertedCAPIMachine)
 
-		return ctrl.Result{}, nil
+	// Make a deep copy of the converted CAPI machine to avoid modifying the original.
+	updatedOrCreatedCAPIMachine := convertedCAPIMachine.DeepCopy()
```

Comment:
This will never be the `created` CAPI machine, because we return in case of creation already above.

Intended?